### PR TITLE
[SQL] Support for user-defined structure types

### PR DESCRIPTION
**Pull request by mihaibudiu**
Is this a user-visible change (yes/no): yes

### CODE REVIEW

**Comment by mihaibudiu** [comment]
@abhizer you may not be able to read most of this code, but we'll have to add lots of tests related to structs
For one, I have noticed that equality comparisons with structs crash calcite

**Comment by gz** [comment]
This will need some support in the UI at least for parsing the schema info that contains the structured types and displaying it correctly @Karakatiza666

**Comment by mihaibudiu** [comment]
@gz: the Rust program schema cannot currently represent record types. Can you help implement the required part as a commit to this PR? Here is an example program and it's JSON schema.

```sql
            CREATE TYPE address_typ AS (
               street          VARCHAR(30),
               city            VARCHAR(30),
               state           CHAR(2),
               postal_code     VARCHAR(6));
            CREATE TYPE person_typ AS (
               firstname       VARCHAR(30),
               lastname        VARCHAR(30),
               address         ADDRESS_TYP);
            CREATE TABLE PERS(p0 PERSON_TYP, p1 PERSON_TYP);
```

**Comment by mihaibudiu** [comment]
```json
{
  "name": "PERS",
  "case_sensitive": false,
  "fields": [
    {
      "name": "P0",
      "case_sensitive": false,
      "columntype": {
        "fields": [
          {
            "type": "VARCHAR",
            "nullable": true,
            "precision": 30,
            "name": "FIRSTNAME"
          },
          {
            "type": "VARCHAR",
            "nullable": true,
            "precision": 30,
            "name": "LASTNAME"
          },
          {
            "fields": {
              "fields": [
                {
                  "type": "VARCHAR",
                  "nullable": true,
                  "precision": 30,
                  "name": "STREET"
                },
                {
                  "type": "VARCHAR",
                  "nullable": true,
                  "precision": 30,
                  "name": "CITY"
                },
                {
                  "type": "CHAR",
                  "nullable": true,
                  "precision": 2,
                  "name": "STATE"
                },
                {
                  "type": "VARCHAR",
                  "nullable": true,
                  "precision": 6,
                  "name": "POSTAL_CODE"
                }
              ],
              "nullable": false
            },
            "nullable": false,
            "name": "ADDRESS"
          }
        ],
        "nullable": false
      }
    },
    {
      "name": "P1",
      "case_sensitive": false,
      "columntype": {
        "fields": [
          {
            "type": "VARCHAR",
            "nullable": true,
            "precision": 30,
            "name": "FIRSTNAME"
          },
          {
            "type": "VARCHAR",
            "nullable": true,
            "precision": 30,
            "name": "LASTNAME"
          },
          {
            "fields": {
              "fields": [
                {
                  "type": "VARCHAR",
                  "nullable": true,
                  "precision": 30,
                  "name": "STREET"
                },
                {
                  "type": "VARCHAR",
                  "nullable": true,
                  "precision": 30,
                  "name": "CITY"
                },
                {
                  "type": "CHAR",
                  "nullable": true,
                  "precision": 2,
                  "name": "STATE"
                },
                {
                  "type": "VARCHAR",
                  "nullable": true,
                  "precision": 6,
                  "name": "POSTAL_CODE"
                }
              ],
              "nullable": false
            },
            "nullable": false,
            "name": "ADDRESS"
          }
        ],
        "nullable": false
      }
    }
  ]
}
```

**Comment by gz** [comment]
@ryzhyk should review my changes in the adapters crate

**Review by ryzhyk — APPROVED** [review]
Only read the Rust code.

File: crates/adapters/src/format/parquet/mod.rs [hunk 1]
```diff
@@ -246,20 +261,21 @@ fn relation_to_parquet_schema(relation: &Relation) -> Result<SerdeArrowSchema, C
             SqlType::Null => DataType::Null,
             SqlType::Binary | SqlType::Varbinary | SqlType::Interval => todo!(),
             SqlType::Array => todo!("handle array types"),
```
Comment (ryzhyk) [inline crates/adapters/src/format/parquet/mod.rs:263]: actually, would be nice to convert this into an error, since this can be triggered by user input

Comment (gz) [reply crates/adapters/src/format/parquet/mod.rs:263]: Or I should just implement this

Comment (ryzhyk) [reply crates/adapters/src/format/parquet/mod.rs:263]: How do you represent arrays in parquet?

Comment (gz) [reply crates/adapters/src/format/parquet/mod.rs:263]:
```
SqlType::Array => {
                // SqlType::Array implies c.component.is_some()
                let array_component = c.component.as_ref().unwrap();
                DataType::LargeList(Arc::new(ArrowField::new_list_field(
                    columntype_to_datatype(array_component),
                    c.nullable,
                )))
            }
```

**Pull request merged**